libkernelflinger/trusty_efi.c

```diff
@@ -181,17 +188,20 @@ static EFI_STATUS get_address_size_vmm(OUT UINT64 *vmm_mem_base, OUT UINT32 *vmm
         if (!vmm_mem_base || !vmm_size)
                 return EFI_INVALID_PARAMETER;
 
-        *vmm_mem_base = VMM_MEM_BASE;
+        *vmm_mem_base = BASE_4GB; //allocate vmm's mem under 4G
         *vmm_size = VMM_MEM_SIZE;
 
         ret = allocate_pages(AllocateMaxAddress,
                              EfiRuntimeServicesData,
-                             EFI_SIZE_TO_PAGES(VMM_MEM_SIZE),
+                             EFI_SIZE_TO_PAGES(VMM_MEM_SIZE + SIZE_2MB), //allocate additional 2MB for alignment
                              vmm_mem_base);
         if (EFI_ERROR(ret)) {
                 efi_perror(ret, L"Alloc memory for VMM base addess failed");
                 return EFI_OUT_OF_RESOURCES;
         }
+
+	*vmm_mem_base = ALIGN(*vmm_mem_base, SIZE_2MB);
```

Comment:
is this ALIGN aligen force or back?

Reply:
ALIGN() is align forward, there is another macro ALIGN_DOWN() to align backward.
You can check the code here: https://github.com/intel/kernelflinger/blob/daee746b1cdc7e2eff1acf1a7d1d9ecaa9dfb86e/include/libkernelflinger/uefi_utils.h#L41-L43